services/src/main/java/org/keycloak/authentication/DefaultAuthenticationFlow.java

```diff
@@ -430,16 +430,17 @@ private Response processSingleFlowExecutionModel(AuthenticationExecutionModel mo
             if (authUser == null) {
                 throw new AuthenticationFlowException("authenticator: " + factory.getId(), AuthenticationFlowError.UNKNOWN_USER);
             }
-            if (!authenticator.configuredFor(processor.getSession(), processor.getRealm(), authUser)) {
-                if (factory.isUserSetupAllowed() && model.isRequired() && authenticator.areRequiredActionsEnabled(processor.getSession(), processor.getRealm())) {
-                    //This means that having even though the user didn't validate the
-                    logger.debugv("authenticator SETUP_REQUIRED: {0}", factory.getId());
-                    processor.getAuthenticationSession().setExecutionStatus(model.getId(), AuthenticationSessionModel.ExecutionStatus.SETUP_REQUIRED);
-                    authenticator.setRequiredActions(processor.getSession(), processor.getRealm(), processor.getAuthenticationSession().getAuthenticatedUser());
-                    return null;
-                } else {
-                    throw new AuthenticationFlowException("authenticator: " + factory.getId(), AuthenticationFlowError.CREDENTIAL_SETUP_REQUIRED);
-                }
+        }
+        //  Trigger authenticator setup if needed if possible and if allowed
+        if (authUser != null && factory.isUserSetupAllowed() && !authenticator.configuredFor(processor.getSession(), processor.getRealm(), authUser)) {
```

Comment:
It seems that L435 and L436 changes the original codes' logic for condition of throwing `AuthenticationFlowException` in L442.

To keep the same condition for throwing it in the original codes, how about the following?

```
    if (authUser != null && !authenticator.configuredFor(processor.getSession(), processor.getRealm(), authUser)) {
        if (factory.isUserSetupAllowed() && model.isRequired() && authenticator.areRequiredActionsEnabled(processor.getSession(), processor.getRealm())) {
```

If you intentionally change this condition, could you tell me this reason?

Reply:
The purpose of the modification is to handle credential setup for authenticators that don't require a user to be set. Your proposition looks good to me, I'll update the code accordingly in order to stay the closest possible to the existing code (though I think we should get rid of the meaningless comment line)

Reply:
@vanrar68 Thank you. I've checked your update.